play/src/front/Phaser/ECS/Entity.ts

```diff
@@ -0,0 +1,287 @@
+    }
+
+    public deactivate(): void {
+        actionsMenuStore.clear();
+    }
+
+    public getCollisionGrid(): number[][] | undefined {
+        return this.entityData.prefab.collisionGrid;
+    }
+
+    public getReversedCollisionGrid(): number[][] | undefined {
+        return this.entityData.prefab.collisionGrid?.map((row) => row.map((value) => (value === 1 ? -1 : value)));
+    }
+
+    public setFollowOutlineColor(color: number): void {
+        this.outlineColorStore.setFollowColor(color);
+    }
+
+    public removeFollowOutlineColor(): void {
+        this.outlineColorStore.removeFollowColor();
+    }
+
+    public setApiOutlineColor(color: number): void {
+        this.outlineColorStore.setApiColor(color);
+    }
+
+    public removeApiOutlineColor(): void {
+        this.outlineColorStore.removeApiColor();
+    }
+
+    public pointerOverOutline(color: number): void {
+        this.outlineColorStore.pointerOver(color);
+    }
+
+    public pointerOutOutline(): void {
+        this.outlineColorStore.pointerOut();
+    }
+
+    public characterCloseByOutline(color: number): void {
+        this.outlineColorStore.characterCloseBy(color);
+    }
+
+    public characterFarAwayOutline(): void {
+        this.outlineColorStore.characterFarAway();
+    }
+
+    public delete() {
+        this.emit(EntityEvent.Remove);
+    }
+
+    private getOutlinePlugin(): OutlinePipelinePlugin | undefined {
+        return this.scene.plugins.get("rexOutlinePipeline") as unknown as OutlinePipelinePlugin | undefined;
+    }
+
+    private hasAnyPropertiesSet(): boolean {
+        if (!this.entityData.properties) {
+            return false;
+        }
+        const propValues = Object.values(this.entityData.properties);
+        for (const value of propValues) {
+            if (value !== undefined && value !== null) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private toggleActionsMenu(): void {
+        if (get(actionsMenuStore) !== undefined) {
+            actionsMenuStore.clear();
+            return;
+        }
+        actionsMenuStore.initialize(isTextHeaderPropertyData.parse(this.entityData.properties["textHeader"] ?? ""));
+        for (const action of this.getDefaultActionsMenuActions()) {
+            actionsMenuStore.addAction(action);
+        }
+    }
+
+    private getDefaultActionsMenuActions(): ActionsMenuAction[] {
+        if (!this.entityData.properties) {
+            return [];
+        }
+        const actions: ActionsMenuAction[] = [];
+        for (const key of Object.keys(this.entityData.properties)) {
+            if (this.entityData.properties[key]) {
+                switch (key) {
+                    case "textHeader": //do nothing, handled in toggleActionsMenu
+                        break;
+                    case "jitsiRoom": {
+                        const propertyData = isJitsiRoomPropertyData.parse(this.entityData.properties[key]);
```

Comment:
You can do even better with Zod by building a Zod type for the "properties" object.

Rather than applying Zod to sub-parts of the JSON object, you can actually apply it at the top level object.

So ideally, you should have a Zod type that defines an array of entities. Each entity is a Zod type itself that contains an array of properties.